### PR TITLE
chore(RWA/RNA): deprecate passwordSettings prop

### PR DESCRIPTION
**Pull request by calebpollman**
#### Description of changes

The `passwordSettings` prop of the `Authenticator` is undocumented and unused in the React and React Native implementations and will be removed in the next major version release of both components.

- migrate existing `useDeprecationWarning` to react-core
- add React & RN `useDeprecationWarning`
- add deprecation warning to `passwordSettings` prop of `Authenticator`

>  Vue and Angular are excluded  as `passwordSettings` is not a configurable property in either implementation

#### Issue #, if available
Related https://github.com/aws-amplify/amplify-ui/issues/4162

#### Description of how you validated changes
- added unit tests for new and existing code
- manually tested console warnings in React and React Native
 
#### Checklist

- [x] Have read the [Pull Request Guidelines](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md)
- [x] PR description included
- [x] `yarn test` passes and tests are updated/added

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by wlee221 — APPROVED** [review]

File: packages/react-core/src/hooks/useDeprecationWarning.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+export type UseDeprecationWarning = (params: {
+  shouldWarn: boolean;
+  message: string;
+}) => void;
+
+/**
+ * Logs a deprecation warning message.
+ *
+ * @inportant Please use the React/React Native specific platform implementations as the
+ * hook and does not consider take whether the code is running in development or prod.
```
Comment (calebpollman) [inline packages/react-core/src/hooks/useDeprecationWarning.ts:12]:
```suggestion
 * @important Please use the React/React Native specific platform implementations.
 * This version of the hook is a base implementation that the others extend from due
 * to env differences between running in RN or the browser
 ```

**Review by zchenwei — APPROVED** [review]
LGTM. Thanks for tackling this! @calebpollman

**Review by wlee221 — APPROVED** [review]

File: packages/react-core/src/hooks/useDeprecationWarning.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ * to env differences between running in RN or the browser
+ */
+const useDeprecationWarning: UseDeprecationWarning = ({
+  shouldWarn,
+  message,
+}) => {
+  React.useEffect(() => {
+    if (shouldWarn) {
+      // eslint-disable-next-line no-console
+      console.warn(message);
```
Comment (ioanabrooks) [inline packages/react-core/src/hooks/useDeprecationWarning.ts:22]: Could we use Logger from `aws-amplify` library and not have to eslint-disable? Thinking it should have the same outcome.

Comment (calebpollman) [reply packages/react-core/src/hooks/useDeprecationWarning.ts:22]: Discussed offline

**Pull request merged**